lib/async_hooks.js

```diff
@@ -264,8 +262,8 @@ class AsyncLocalStorage {
     }
   }
 
-  runSyncAndReturn(callback, ...args) {
-    this._enter();
+  runSyncAndReturn(store, callback, ...args) {
```

Comment:
Should we disallow `null` and `undefined` here?

Reply:
Is there a specific _reason_ to?

Reply:
`null` or `undefined` values for `store` shouldn't be a problem - if users want it as the store, they get it. As for the `callback`, the subsequent invocation will throw. So, I'm not sure if we need any extra checks.

Reply:
This could be somewhat hot code, and I don't think there's any significant _reason_ to _prevent_ using `null` or `undefined` here, so I think the best thing is to just do nothing.

Reply:
If you use `undefined` as store you get `undefined` if you call `getStore()` if you are within the async tree but you get `undefined` also if you are not on context.
Therefore I think this is always wrong and adding a check would help users to avoid such mistakes.
Regarding hot path: I think it's much more expensive if we start propagating `undefined` via async hooks compared to adding an if here.

Reply:
So, personally, I am very opinionated about what a store should be (I know it's not a shared opinion). I expect the store to be`typeof store === 'object'` (no functions, nor `undefined`).

I believe we should enforce as many best practices we can in this API. Being too permissive in term of store actually pushes users to use multiple instances of the API (one per concern) - which comes with a performance and complexity costs. In an ideal world, there would be only a few instances of AsyncLocalStorage in an app. One for APM/RASP, one for the framework, one for the user's code. No more. For instance, instanciating an AsyncLocalStorage just to keep a request id, is a terribly bad practice.

However, it seems there is a consencus around the current change and I am certain I can add enough warnings in litterature (doc, guides and blogposts) to highlights the risks/impacts. That's pretty much why I am ok-ish with not enforcing enforcing anything regarding the nature of the store.

Reply:
I'm fine with docs discouraging the use of exotic store types, but I'd prefer not to add too much extra type-checking stuff unnecessarily. There could be valid use cases for using those exotic types which we haven't thought of yet, though I don't have strong feeling either way on allowing/disallowing them. If we think there's an actual _safety_ concern that should drive us to control the types more, I'm on board with that. If we're just banning types because of our own opinions of how people should write their code, I'm not into that.

Reply:
@Flarna do you think this is something we can iterate upon later? (we will have a few months of experimental here). I'd like to merge this not too late so we can land AsyncLocalStorage in the next release(s)

Reply:
Sure, don't want to block this. I still think that using `undefined` is a source of hard to find errors but we are using javascript here where this is quite common anyway :grin: